File: [1]Programmers/Level2/다리를지나는트럭.cpp
#include <iostream>
#include <queue>
#include <string>
#include <vector>
using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
  int time = 0;
  int check = 0;
  queue<int> bridge;

  for(int i = 0 ; i < truck_weights.size() ; i++){
    int truck = truck_weights.at(i);

    while(true){
      if(bridge.empty()){
        bridge.push(truck);
        check += truck;
        time++;
        break;
      }else if(bridge.size() == bridge_length){
        check -= bridge.front();
        bridge.pop();
      }else{
        if(check + truck <= weight){
          check += truck;
          bridge.push(truck);
          time++;
          break;
        }else{
          bridge.push(0);
          time++;
        }
      }
    }
  }
  return time + bridge_length;
}

int main(){
  int bridge_length_1 = 2;
  int weight_1 = 10;
  int bridge_length_2 = 100;
  int weight_2 = 100;
  vector<int> truck_weights_1;
  vector<int> truck_weights_2;

  truck_weights_1.push_back(7);
  truck_weights_1.push_back(4);
  truck_weights_1.push_back(5);
  truck_weights_1.push_back(6);
  truck_weights_2.push_back(10);

  cout << solution(bridge_length_1, weight_1, truck_weights_1) << endl;
  cout << solution(bridge_length_2, weight_2, truck_weights_2) << endl;

  return 0;
}

File: JavaScript/[1]Programmers/코딩테스트 고득점 Kit/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {

    let time = 0;
    let onBridge = [];
    let truckSum = 0;

    while (truck_weights.length > 0 || onBridge.length > 0) {
        time++;        
        
        if (onBridge.length > 0) {
            if (time-onBridge[0].time == bridge_length) {
                truckSum -= onBridge[0].weight;
                onBridge.shift();
            }
        }

        if (truck_weights.length > 0) {
            if (truckSum+truck_weights[0] <= weight) {
                truckSum += truck_weights[0];
                onBridge.push({ weight : truck_weights.shift(), time : time});
            }
        }
    }        
    return time;
}

let bridge_length1 = 2;
let weight1 = 10;
let truck_weights1 = [7,4,5,6];

let bridge_length2 = 100;
let weight2 = 100;
let truck_weights2 = [10];

let bridge_length3 = 100;
let weight3 = 100;
let truck_weights3 = [10,10,10,10,10,10,10,10,10,10];

console.log(solution(bridge_length1, weight1, truck_weights1));